testing.rst

```diff
@@ -11,377 +11,513 @@ using both functional and unit tests.
+used. If you have learned something different, that is not necessarily
+wrong, just different from what the Symfony documentation is using.
 
-.. tip::
+`Unit Tests`_
+    These tests ensure that *individual* units of source code (e.g. a single
+    class) behave as intended.
 
-    Use the ``--coverage-*`` command options to generate code coverage reports.
-    Read the PHPUnit manual to learn more about `code coverage analysis`_.
+`Integration Tests`_
+    These tests test a combination of classes and commonly interact with
+    Symfony's service container. These tests do not yet cover the fully
+    working application, those are called *Application tests*.
 
-.. index::
-   single: Tests; Unit tests
+`Application Tests`_
+    Application tests test the behavior of a complete application. They
+    make, sometimes fake, HTTP requests and test that the response is as
+    expected.
 
 Unit Tests
 ----------
 
-A `unit test`_ ensures that individual units of source code (e.g. a single class
-or some specific method in some class) meet their design and behave as intended.
-If you want to test an entire feature of your application (e.g. registering as a
-user or generating an invoice), see the section about :ref:`Functional Tests <functional-tests>`.
+A `unit test`_ ensures that individual units of source code (e.g. a single
+class or some specific method in some class) meet their design and behave
+as intended. Writing unit tests in a Symfony application no different from
+writing standard PHPUnit unit tests. You can learn about it in the PHPUnit
+documentation: `Writing Tests for PHPUnit`_.
 
-Writing Symfony unit tests is no different from writing standard PHPUnit
-unit tests. Suppose, for example, that you have an class called ``Calculator``
-in the ``src/Util/`` directory of the app::
+By convention, the ``tests/`` directory should replicate the directory
+of your application for unit tests. So, if you're testing a class in the
+``src/Util/`` directory, put the test in the ``tests/Util/`` directory.
+Autoloading is automatically enabled via the ``vendor/autoload.php`` file
+(as configured by default in the ``phpunit.xml.dist`` file).
 
-    // src/Util/Calculator.php
-    namespace App\Util;
+You can run tests using the ``./vendor/bin/phpunit`` command:
 
-    class Calculator
-    {
-        public function add($a, $b)
-        {
-            return $a + $b;
-        }
-    }
+.. code-block:: terminal
+
+    # run all tests of the application
+    $ php ./vendor/bin/phpunit
+
+    # run all tests in the Util/ directory
+    $ php ./vendor/bin/phpunit tests/Util
+
+    # run tests for the Calculator class
+    $ php ./vendor/bin/phpunit tests/Util/CalculatorTest.php
+
+.. tip::
+
+    In large test suites, it can make sense to create subdirectories for
+    each type of tests (e.g. ``tests/unit/`` and ``test/functional/``). In
+    this case, it is recommended that each subdirectory replicates the
+    application directory structure.
```

Comment:
Hm.. Im not sure if we should mention this. Sure it is valid, but it has little real benefit. If you are using PHPUnit test suites, then this has zero benefits.